Miscompilation of certain code

### Description

```nim
proc retInt(): int {.discardable.} =
    42

proc retString(): string {.discardable.} =
    "text"

proc doStuff(msg: string) =
    case msg:
    of "a":
        retString()
    of "b":
        retInt()
    of "c":
        let _ = retInt()
    else:
        raise newException(ValueError, "error")

doStuff("q")

```

### Nim Version

Nim Compiler Version 2.0.6 [MacOSX: arm64]
Compiled at 2024-06-17

### Current Output

```text
Info:  Nimble data file "/Users/_/.nimble/nimbledata2.json" has been loaded.
  Verifying dependencies for test_app@0.1.0
   Building test_app/test_app using c backend
  Executing /opt/homebrew/Cellar/nim/2.0.6/nim/bin/nim c --colors:on --noNimblePath -d:NimblePkgVersion=0.1.0 -o:/Users/_/test_app/test_app /Users/_/test_app/src/test_app.nim
Hint: used config file '/opt/homebrew/Cellar/nim/2.0.6/nim/config/nim.cfg' [Conf]
Hint: used config file '/opt/homebrew/Cellar/nim/2.0.6/nim/config/config.nims' [Conf]
......................................................................
CC: test_app.nim
/Users/_/.cache/nim/test_app_d/@mtest_app.nim.c:234:14: error: assigning to 'NimStringV2' (aka 'struct NimStringV2') from incompatible type 'NI' (aka 'long long')
                colontmpD_ = colontmpD__2;
                           ^ ~~~~~~~~~~~~
1 error generated.
Error: execution of an external compiler program 'clang -c  -w -ferror-limit=3 -pthread   -I/opt/homebrew/Cellar/nim/2.0.6/nim/lib -I/Users/_/test_app/src -o /Users/_/.cache/nim/test_app_d/@mtest_app.nim.c.o /Users/_/.cache/nim/test_app_d/@mtest_app.nim.c' failed with exit code: 1


nimble.nim(304)          buildFromDir

    Error:  Build failed for the package: test_app
     Info:  Nimble data file "/Users/_/.nimble/nimbledata2.json" has been saved.
```

### Expected Output

### Possible Solution

As far as I understand, nim doesn't like when different `case` branches return different value types, even when those values are unused. Usually it simply returns a "type mismatch" error, which is annoying, but a separate problem. But here it generates C code just fine, but fails at compiling it.

By removing `{.discardable.}` or by using `let _ = ...` in branches we can "fix" this issue.

### Additional Information

--- a/compiler/injectdestructors.nim
+++ b/compiler/injectdestructors.nim
@@ -653,7 +653,7 @@ template handleNestedTempl(n, processCall: untyped, willProduceStmt = false,
       for j in 0 ..< it.len-1:
         branch[j] = copyTree(it[j])
       var ofScope = nestedScope(s, it.lastSon)
-      branch[^1] = if it[^1].typ.isEmptyType or willProduceStmt:
+      branch[^1] = if n.typ.isEmptyType or willProduceStmt:
                      processScope(c, ofScope, maybeVoid(it[^1], ofScope))
                    else:
                      processScopeExpr(c, ofScope, it[^1], processCall, tmpFlags)
@@ -701,7 +701,7 @@ template handleNestedTempl(n, processCall: untyped, willProduceStmt = false,
         #Condition needs to be destroyed outside of the condition/branch scope
         branch[0] = p(it[0], c, s, normal)
 
-      branch[^1] = if it[^1].typ.isEmptyType or willProduceStmt:
+      branch[^1] = if n.typ.isEmptyType or willProduceStmt:
                      processScope(c, branchScope, maybeVoid(it[^1], branchScope))
                    else:
                      processScopeExpr(c, branchScope, it[^1], processCall, tmpFlags)

--- a/compiler/injectdestructors.nim
+++ b/compiler/injectdestructors.nim
@@ -653,7 +653,7 @@ template handleNestedTempl(n, processCall: untyped, willProduceStmt = false,
       for j in 0 ..< it.len-1:
         branch[j] = copyTree(it[j])
       var ofScope = nestedScope(s, it.lastSon)
-      branch[^1] = if n.typ.isEmptyType or willProduceStmt:
+      branch[^1] = if n.typ.isEmptyType or it[^1].typ.isEmptyType or willProduceStmt:
                      processScope(c, ofScope, maybeVoid(it[^1], ofScope))
                    else:
                      processScopeExpr(c, ofScope, it[^1], processCall, tmpFlags)
@@ -701,7 +701,7 @@ template handleNestedTempl(n, processCall: untyped, willProduceStmt = false,
         #Condition needs to be destroyed outside of the condition/branch scope
         branch[0] = p(it[0], c, s, normal)
 
-      branch[^1] = if n.typ.isEmptyType or willProduceStmt:
+      branch[^1] = if n.typ.isEmptyType or it[^1].typ.isEmptyType or willProduceStmt:
                      processScope(c, branchScope, maybeVoid(it[^1], branchScope))
                    else:
                      processScopeExpr(c, branchScope, it[^1], processCall, tmpFlags)

--- a/compiler/semstmts.nim
+++ b/compiler/semstmts.nim
@@ -283,6 +283,7 @@ proc discardCheck(c: PContext, result: PNode, flags: TExprFlags) =
     if implicitlyDiscardable(result):
       var n = newNodeI(nkDiscardStmt, result.info, 1)
       n[0] = result
+      # notes that it doesn't transform nodes into discard statements
     elif result.typ.kind != tyError and c.config.cmd != cmdInteractive:
       if result.typ.kind == tyNone:
         localError(c.config, result.info, "expression has no type: " &

--- a/tests/discard/tdiscardable.nim
+++ b/tests/discard/tdiscardable.nim
@@ -140,3 +140,27 @@ block: # issue #14665
         continue
       inc i
   test()
+
+block: # bug #23775
+  proc retInt(): int {.discardable.} =
+    42
+
+  proc retString(): string {.discardable.} =
+    "text"
+
+  type
+    Enum = enum
+      A, B, C, D
+
+  proc doStuff(msg: Enum) =
+    case msg:
+    of A:
+      retString()
+    of B:
+      retInt()
+    of C:
+      discard retString()
+    else:
+      let _ = retString()
+
+  doStuff(C)

